File: apps/ios/Tests/CrossyProtocolTests/WirePostureTests.swift
import Foundation
import XCTest

import CrossyProtocol

// Parse posture (PROTOCOL.md §3, §5, §11, §14), twinned from the posture cases in
// packages/protocol/src/codec.test.ts: unknown fields are ignored, a
// recognizable-but-unknown `type` is a distinct outcome from a malformed frame, and
// version negotiation is business logic, not decoding.

final class WirePostureTests: XCTestCase {
    func test_unknownClientCommandTypeThrowsUnknownType() throws {
        // §5: the server answers UNKNOWN_TYPE; the decode outcome must be
        // distinguishable from malformed so it can.
        let frame = Data(#"{"type":"frobnicate","commandId":"c9"}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ClientMessage.self, from: frame)) { error in
            XCTAssertEqual(error as? WireDecodingError, .unknownType("frobnicate"))
        }
    }

    func test_unknownServerNoticeTypeThrowsUnknownType() throws {
        // §3: the client ignores and logs an unknown notice; it needs the same
        // distinguishable outcome.
        let frame = Data(#"{"type":"sparkle","glitter":true}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ServerMessage.self, from: frame)) { error in
            XCTAssertEqual(error as? WireDecodingError, .unknownType("sparkle"))
        }
    }

    func test_unknownFieldsAreIgnoredAndDroppedOnReencode() throws {
        // §3, §14 forward compatibility: decode copies only known fields, exactly as
        // the TS decoders build a fresh object.
        let frame = Data(
            #"""
            {"type":"cellSet","seq":1,"cell":0,"value":"A","by":"u1",
             "commandId":"c1","at":"2026-07-07T00:00:00Z","futureField":{"nested":true}}
            """#.utf8)
        let decoded = try JSONDecoder().decode(ServerMessage.self, from: frame)
        let reencoded = try XCTUnwrap(
            try jsonObject(JSONEncoder().encode(decoded)) as? NSDictionary)
        XCTAssertNil(reencoded["futureField"])
        XCTAssertEqual(
            Set(reencoded.allKeys.compactMap { $0 as? String }),
            ["type", "seq", "cell", "value", "by", "commandId", "at"])
    }

    func test_malformedFramesThrowDecodingErrorNotUnknownType() throws {
        // §11: a frame with no usable `type` is malformed (drop-and-log posture), a
        // different failure from unknown_type. Mirrors the TS malformed cases.
        let malformed = ["42", #""string""#, "null", "[]", "{}", #"{"type":7}"#]
        for raw in malformed {
            XCTAssertThrowsError(
                try JSONDecoder().decode(ClientMessage.self, from: Data(raw.utf8)),
                "frame \(raw) must fail to decode"
            ) { error in
                XCTAssertTrue(
                    error is DecodingError,
                    "frame \(raw) must be malformed (DecodingError), got \(error)")
            }
        }
    }

    func test_missingRequiredFieldIsMalformed() throws {
        let frame = Data(#"{"type":"placeLetter","commandId":"c1","cell":0}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ClientMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError, "missing `value` must be malformed")
        }
    }

    func test_cellSetValueKeyIsRequiredEvenThoughNullable() throws {
        // §6: `value` is nullable-and-present. A cellSet with no `value` key at all is
        // malformed, matching the TS asNullableString, not silently nil.
        let frame = Data(
            #"{"type":"cellSet","seq":1,"cell":0,"by":"u1","commandId":"c1","at":"t"}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ServerMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError)
        }
    }

    func test_helloForAFutureVersionStillDecodes() throws {
        // §2, §14: any integer protocolVersion decodes cleanly; mapping an unsupported
        // one to PROTOCOL_VERSION_UNSUPPORTED is the server's negotiation, not the codec's.
        let frame = Data(#"{"type":"hello","protocolVersion":999,"token":"jwt"}"#.utf8)
        let decoded = try JSONDecoder().decode(ClientMessage.self, from: frame)
        guard case .hello(let hello) = decoded else {
            return XCTFail("expected a hello, got \(decoded.type)")
        }
        XCTAssertEqual(hello.protocolVersion, 999)
    }

    func test_unknownErrorCodeIsMalformed() throws {
        // §11 twin of the TS asErrorCode: the code vocabulary is closed per protocol
        // version (a new code is a §14 concern), so "NONSENSE" is malformed.
        let frame = Data(#"{"type":"error","code":"NONSENSE","message":"x","fatal":false}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ServerMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError)
        }
    }

    func test_asciiOnlyWireEnums_INV1() throws {
        // INV-1: every enum raw value this module compares against the wire is plain
        // ASCII, so no locale-aware transform can be involved in matching them.
        let rawValues =
            Role.allCases.map(\.rawValue)
            + Direction.allCases.map(\.rawValue)
            + GameStatus.allCases.map(\.rawValue)
            + ErrorCode.allCases.map(\.rawValue)
            + APIErrorCode.allCases.map(\.rawValue)
        for raw in rawValues {
            XCTAssertTrue(raw.utf8.allSatisfy { $0 < 0x80 }, "\(raw) must be ASCII")
        }
    }
}

File: apps/android/protocol/src/test/kotlin/crossy/protocol/WirePostureTests.kt
package crossy.protocol

import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.elementNames
import kotlinx.serialization.json.jsonObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

// Parse posture (PROTOCOL.md §3, §5, §11, §14), twinned from the posture cases in
// packages/protocol/src/codec.test.ts and apps/ios WirePostureTests.swift: unknown fields are
// ignored, a recognizable-but-unknown `type` is a distinct outcome from a malformed frame, and
// version negotiation is business logic, not decoding.

class WirePostureTests {
    @Test
    fun unknownClientCommandTypeThrowsUnknownType() {
        // §5: the server answers UNKNOWN_TYPE; the decode outcome must be distinguishable from
        // malformed so it can.
        val frame = """{"type":"frobnicate","commandId":"c9"}"""
        val error = assertThrows<WireDecodingException.UnknownType> {
            ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        }
        assertEquals(WireDecodingException.UnknownType("frobnicate"), error)
    }

    @Test
    fun unknownServerNoticeTypeThrowsUnknownType() {
        // §3: the client ignores and logs an unknown notice; it needs the same distinguishable outcome.
        val frame = """{"type":"sparkle","glitter":true}"""
        val error = assertThrows<WireDecodingException.UnknownType> {
            ProtocolJson.decodeFromString(ServerMessageSerializer, frame)
        }
        assertEquals(WireDecodingException.UnknownType("sparkle"), error)
    }

    @Test
    fun unknownFieldsAreIgnoredAndDroppedOnReencode() {
        // §3, §14 forward compatibility: decode copies only known fields, exactly as the TS
        // decoders build a fresh object.
        val frame = """{"type":"cellSet","seq":1,"cell":0,"value":"A","by":"u1","commandId":"c1","at":"2026-07-07T00:00:00Z","futureField":{"nested":true}}"""
        val decoded = ProtocolJson.decodeFromString(ServerMessageSerializer, frame)
        val reencoded = ProtocolJson.parseToJsonElement(
            ProtocolJson.encodeToString(ServerMessageSerializer, decoded),
        ).jsonObject
        assertFalse(reencoded.containsKey("futureField"))
        assertEquals(setOf("type", "seq", "cell", "value", "by", "commandId", "at"), reencoded.keys)
    }

    @Test
    fun malformedFramesThrowSerializationExceptionNotUnknownType() {
        // §11: a frame with no usable `type` is malformed (drop-and-log posture), a different
        // failure from unknown_type. Mirrors the TS malformed cases.
        val malformed = listOf("42", "\"string\"", "null", "[]", "{}", """{"type":7}""")
        for (raw in malformed) {
            val error = assertThrows<SerializationException>("frame $raw must fail to decode") {
                ProtocolJson.decodeFromString(ClientMessageSerializer, raw)
            }
            assertFalse(error is WireDecodingException, "frame $raw must be malformed, not unknown-type")
        }
    }

    @Test
    fun missingRequiredFieldIsMalformed() {
        val frame = """{"type":"placeLetter","commandId":"c1","cell":0}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        }
    }

    @Test
    fun cellSetValueKeyIsRequiredEvenThoughNullable() {
        // §6: `value` is nullable-and-present. A cellSet with no `value` key at all is malformed,
        // matching the TS asNullableString, not silently null.
        val frame = """{"type":"cellSet","seq":1,"cell":0,"by":"u1","commandId":"c1","at":"t"}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ServerMessageSerializer, frame)
        }
    }

    @Test
    fun helloForAFutureVersionStillDecodes() {
        // §2, §14: any integer protocolVersion decodes cleanly; mapping an unsupported one to
        // PROTOCOL_VERSION_UNSUPPORTED is the server's negotiation, not the codec's.
        val frame = """{"type":"hello","protocolVersion":999,"token":"jwt"}"""
        val decoded = ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        assertTrue(decoded is ClientMessage.Hello, "expected a hello, got ${decoded.type}")
        assertEquals(999, (decoded as ClientMessage.Hello).message.protocolVersion)
    }

    @Test
    fun unknownErrorCodeIsMalformed() {
        // §11 twin of the TS asErrorCode: the code vocabulary is closed per protocol version (a new
        // code is a §14 concern), so "NONSENSE" is malformed.
        val frame = """{"type":"error","code":"NONSENSE","message":"x","fatal":false}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ServerMessageSerializer, frame)
        }
    }

    @Test
    fun asciiOnlyWireEnums_INV1() {
        // INV-1: every enum wire value this module compares against the wire is plain ASCII, so no
        // locale-aware transform can be involved in matching them.
        val wireValues =
            Role.serializer().descriptor.elementNames +
                Direction.serializer().descriptor.elementNames +
                GameStatus.serializer().descriptor.elementNames +
                ErrorCode.serializer().descriptor.elementNames +
                APIErrorCode.entries.map { it.name }
        for (raw in wireValues) {
            assertTrue(raw.all { it.code < 0x80 }, "$raw must be ASCII")
        }
    }
}
